estimagic/optimization/nlopt_optimizers.py

```diff
@@ -141,6 +140,899 @@ def nlopt_neldermead(
+        stopping_max_eval=stopping_max_criterion_evaluations,
+    )
+
+    return out
+
+
+def nlopt_slsqp(
+    criterion_and_derivative,
+    x,
+    lower_bounds,
+    upper_bounds,
+    *,
+    convergence_relative_params_tolerance=CONVERGENCE_RELATIVE_PARAMS_TOLERANCE,
+    convergence_absolute_params_tolerance=CONVERGENCE_ABSOLUTE_PARAMS_TOLERANCE,
+    convergence_relative_criterion_tolerance=CONVERGENCE_RELATIVE_CRITERION_TOLERANCE,
+    convergence_absolute_criterion_tolerance=CONVERGENCE_ABSOLUTE_CRITERION_TOLERANCE,
+    stopping_max_criterion_evaluations=STOPPING_MAX_CRITERION_EVALUATIONS,
+):
+    """Optimize a scalar function based on SLSQP method.
+
+    SLSQP solves gradient based nonlinearly constrained optimization problems.
+    The algorithm treats the optimization problem as a sequence of constrained
+    least-squares problems.
+
+    The implementation is based on the procedure described in:
+    Dieter Kraft, "A software package for sequential quadratic programming",
+    Technical Report DFVLR-FB 88-28, Institut für Dynamik der Flugsysteme,
+    Oberpfaffenhofen, July 1988.
+    Dieter Kraft, "Algorithm 733: TOMP–Fortran modules for optimal control
+    calculations," ACM Transactions on Mathematical Software, vol. 20, no. 3,
+    pp. 262-281 (1994).
+
+    ``nlopt_slsqp`` supports the following ``algo_options``:
+
+    - convergence.relative_params_tolerance (float):  Stop when the relative movement
+      between parameter vectors is smaller than this.
+    - convergence.relative_criterion_tolerance (float): Stop when the relative
+      improvement between two iterations is smaller than this.
+    - stopping_max_criterion_evaluations (int): If the maximum number of function
+      evaluation is reached, the optimization stops but we do not count this
+      as convergence.
+
+    """
+    out = _minimize_nlopt(
+        criterion_and_derivative,
+        x,
+        lower_bounds,
+        upper_bounds,
+        algorithm=nlopt.LD_SLSQP,
+        algorithm_name="nlopt_slsqp",
+        convergence_xtol_rel=convergence_relative_params_tolerance,
+        convergence_xtol_abs=convergence_absolute_params_tolerance,
+        convergence_ftol_rel=convergence_relative_criterion_tolerance,
+        convergence_ftol_abs=convergence_absolute_criterion_tolerance,
+        stopping_max_eval=stopping_max_criterion_evaluations,
+    )
+    return out
+
+
+def nlopt_direct(
+    criterion_and_derivative,
+    x,
+    lower_bounds,
+    upper_bounds,
+    *,
+    convergence_relative_params_tolerance=CONVERGENCE_RELATIVE_PARAMS_TOLERANCE,
+    convergence_absolute_params_tolerance=CONVERGENCE_ABSOLUTE_PARAMS_TOLERANCE,
+    convergence_relative_criterion_tolerance=CONVERGENCE_RELATIVE_CRITERION_TOLERANCE,
+    convergence_absolute_criterion_tolerance=CONVERGENCE_ABSOLUTE_CRITERION_TOLERANCE,
+    stopping_max_criterion_evaluations=STOPPING_MAX_CRITERION_EVALUATIONS_GLOBAL,
+    nlopt_direct_version=0,
```

Comment:
Since all of those are binary options that can be combined in any way, I would add three boolean arguments instead of the one version argument: `locally_biased`, `unscaled_bounds` and `random`. Then you determine the correct algorithm from that. 

More general: Never ever use integer options or returns that a user has to look up in the documentation. It might be justifiable when using C or Fortran but in Python we always want to automate the lookup.

Reply:
Thanks